File: csharp/MudbaseShowcase.Kanban/Services/MudbaseDataService.cs
using System.Net;
using System.Text.Json;
using Microsoft.Extensions.Options;
using Mudbase.Sdk.Api;
using Mudbase.Sdk.Client;
using Mudbase.Sdk.Model;
using MudbaseShowcase.Kanban.Options;

namespace MudbaseShowcase.Kanban.Services;

/// <summary>A page of documents plus Mudbase's pagination metadata.</summary>
public sealed class MudbaseListResult<T>
{
    public IReadOnlyList<T> Items { get; init; } = Array.Empty<T>();
    public int Page { get; init; }
    public int Limit { get; init; }
    public int Total { get; init; }
    public int TotalPages { get; init; }
}

/// <summary>
/// Thin, generic wrapper over Mudbase.Sdk.Api.DataApi for the lists/cards/activity collections.
/// Handles the plumbing the generated SDK leaves to the caller: building the `filter` query param
/// as a JSON string, and reconstructing a full document from the SDK's untyped response shapes.
///
/// - Single-document reads (Get/Create/Update) come back as `DataResponse.Data` typed `Object`,
///   which System.Text.Json deserializes to a boxed JsonElement - re-serialize/deserialize that
///   into our own POCO.
/// - List reads: the generated SDK's list-item model (`DataListResponseDataInner`) now correctly
///   populates its `AdditionalProperties` dictionary for every field beyond `_id`/`createdAt`/
///   `updatedAt` - an earlier SDK bug that silently discarded every collection-specific field on
///   list reads has been fixed at the SDK level (verified by reading
///   mudbase-sdk/csharp/src/Mudbase.Sdk/Model/DataListResponseDataInner.cs's converter). This
///   service still parses `response.RawContent` directly with a plain <see cref="JsonDocument"/>
///   rather than reconstructing from `AdditionalProperties`, because doing so maps straight to our
///   POCOs in one step and needs no code change if the SDK's internal representation ever changes
///   again - the authenticated HTTP call itself (token attachment, 401-refresh-retry) still goes
///   through the real SDK either way.
/// </summary>
public sealed class MudbaseDataService
{
    private readonly IDataApi _dataApi;
    private readonly MudbaseOptions _options;

    public MudbaseDataService(IDataApi dataApi, IOptions<MudbaseOptions> options)
    {
        _dataApi = dataApi;
        _options = options.Value;
    }

    public async Task<MudbaseListResult<T>> ListAsync<T>(
        string collectionId,
        IReadOnlyDictionary<string, object?>? filter = null,
        string sort = "-createdAt",
        int page = 1,
        int limit = 20,
        CancellationToken cancellationToken = default)
    {
        string? filterJson = filter is { Count: > 0 } ? JsonSerializer.Serialize(filter, MudbaseJson.Options) : null;
        Option<string> filterOption = filterJson is null ? default : new Option<string>(filterJson);

        IListDataApiResponse response = await _dataApi.ListDataAsync(
            _options.ProjectId, collectionId, page, limit, sort, filterOption, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            throw MudbaseApiException.From(response);
        }

        using JsonDocument doc = JsonDocument.Parse(response.RawContent);
        JsonElement root = doc.RootElement;

        List<T> items = new();
        if (root.TryGetProperty("data", out JsonElement dataElement) && dataElement.ValueKind == JsonValueKind.Array)
        {
            foreach (JsonElement itemElement in dataElement.EnumerateArray())
            {
                T? mapped = JsonSerializer.Deserialize<T>(itemElement.GetRawText(), MudbaseJson.Options);
                if (mapped is not null) items.Add(mapped);
            }
        }

        int resolvedPage = page;
        int resolvedLimit = limit;
        int resolvedTotal = items.Count;
        int resolvedTotalPages = 1;
        if (root.TryGetProperty("pagination", out JsonElement paginationElement) && paginationElement.ValueKind == JsonValueKind.Object)
        {
            if (paginationElement.TryGetProperty("page", out JsonElement pageEl)) resolvedPage = pageEl.GetInt32();
            if (paginationElement.TryGetProperty("limit", out JsonElement limitEl)) resolvedLimit = limitEl.GetInt32();
            if (paginationElement.TryGetProperty("total", out JsonElement totalEl)) resolvedTotal = totalEl.GetInt32();
            if (paginationElement.TryGetProperty("totalPages", out JsonElement totalPagesEl)) resolvedTotalPages = totalPagesEl.GetInt32();
        }

        return new MudbaseListResult<T>
        {
            Items = items,
            Page = resolvedPage,
            Limit = resolvedLimit,
            Total = resolvedTotal,
            TotalPages = resolvedTotalPages,
        };
    }

    public async Task<T?> GetAsync<T>(string collectionId, string documentId, CancellationToken cancellationToken = default)
        where T : class
    {
        IGetDataApiResponse response = await _dataApi.GetDataAsync(_options.ProjectId, collectionId, documentId, cancellationToken);

        if (response.StatusCode == HttpStatusCode.NotFound)
        {
            return null;
        }

        if (!response.TryOk(out DataResponse? body) || body?.Data is null)
        {
            throw MudbaseApiException.From(response);
        }

        return MapSingle<T>(body.Data);
    }

    public async Task<T> CreateAsync<T>(string collectionId, IReadOnlyDictionary<string, object?> data, CancellationToken cancellationToken = default)
        where T : class
    {
        ICreateDataApiResponse response = await _dataApi.CreateDataAsync(_options.ProjectId, collectionId, data, cancellationToken);

        if (!response.TryCreated(out DataResponse? body) || body?.Data is null)
        {
            throw MudbaseApiException.From(response);
        }

        return MapSingle<T>(body.Data)!;
    }

    public async Task<T> UpdateAsync<T>(string collectionId, string documentId, IReadOnlyDictionary<string, object?> data, CancellationToken cancellationToken = default)
        where T : class
    {
        IUpdateDataApiResponse response = await _dataApi.UpdateDataAsync(_options.ProjectId, collectionId, documentId, data, cancellationToken);

        if (!response.TryOk(out DataResponse? body) || body?.Data is null)
        {
            throw MudbaseApiException.From(response);
        }

        return MapSingle<T>(body.Data)!;
    }

    public async Task DeleteAsync(string collectionId, string documentId, CancellationToken cancellationToken = default)
    {
        IDeleteDataApiResponse response = await _dataApi.DeleteDataAsync(_options.ProjectId, collectionId, documentId, cancellationToken);

        if (!response.IsOk)
        {
            throw MudbaseApiException.From(response);
        }
    }

    private static T? MapSingle<T>(object data) where T : class
    {
        // System.Text.Json.JsonSerializer.Deserialize<Object> (used internally by the SDK's
        // DataResponseJsonConverter) always produces a boxed JsonElement, never our POCO type
        // directly - re-parse its raw text into the type we actually want.
        if (data is JsonElement element)
        {
            return JsonSerializer.Deserialize<T>(element.GetRawText(), MudbaseJson.Options);
        }

        // Defensive fallback in case a future SDK version deserializes `Object` differently.
        return JsonSerializer.Deserialize<T>(JsonSerializer.Serialize(data, MudbaseJson.Options), MudbaseJson.Options);
    }
}

File: mobile-flutter/lib/core/mudbase_data_service.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:mudbase_sdk/mudbase_sdk.dart';

import 'mudbase_exception.dart';

/// Generic Collections CRUD (`/api/data/projects/:projectId/collections/
/// :collectionId/data`) for the `lists`/`cards`/`activity` collections.
///
/// Bypasses the generated `DataApi` wrapper for the same reason as
/// [AuthService] (see that file's header): `DataApi.listData`'s typed
/// response only declares `_id`/`createdAt`/`updatedAt` - every field this
/// app actually reads (`boardId`, `title`, `position`, `assigneeName`, ...)
/// would be silently dropped from a list read. Ported from the sibling
/// social/ecommerce Flutter apps' `core/mudbase_data_service.dart`.
class MudbaseDataService {
  const MudbaseDataService(this._sdk, this._projectId);

  final MudbaseSdk _sdk;
  final String _projectId;

  String _basePath(String collectionId, [String? documentId]) {
    final base =
        '/api/data/projects/$_projectId/collections/$collectionId/data';
    return documentId == null ? base : '$base/$documentId';
  }

  Options _authOptions(String token, {String? contentType}) {
    return Options(
      contentType: contentType,
      headers: {'Authorization': 'Bearer $token'},
    );
  }

  Future<List<Map<String, dynamic>>> list(
    String collectionId, {
    required String token,
    Map<String, dynamic>? filter,
    String sort = '-createdAt',
    int page = 1,
    int limit = 100,
  }) async {
    final query = <String, dynamic>{
      'page': page,
      'limit': limit,
      'sort': sort,
      if (filter != null && filter.isNotEmpty) 'filter': jsonEncode(filter),
    };
    try {
      final response = await _sdk.dio.get<dynamic>(
        _basePath(collectionId),
        queryParameters: query,
        options: _authOptions(token),
      );
      final body = response.data;
      final items = body is Map<String, dynamic> ? body['data'] : null;
      if (items is! List) return const [];
      return items.whereType<Map<String, dynamic>>().toList(growable: false);
    } on DioException catch (error) {
      throw MudbaseException.fromDioException(error);
    }
  }

  Future<Map<String, dynamic>?> getById(
    String collectionId,
    String documentId, {
    required String token,
  }) async {
    try {
      final response = await _sdk.dio.get<dynamic>(
        _basePath(collectionId, documentId),
        options: _authOptions(token),
      );
      return _unwrapDocument(response);
    } on DioException catch (error) {
      if (error.response?.statusCode == 404) return null;
      throw MudbaseException.fromDioException(error);
    }
  }

  Future<Map<String, dynamic>> create(
    String collectionId,
    Map<String, dynamic> body, {
    required String token,
  }) async {
    try {
      final response = await _sdk.dio.post<dynamic>(
        _basePath(collectionId),
        data: body,
        options: _authOptions(token, contentType: 'application/json'),
      );
      final document = _unwrapDocument(response);
      if (document == null) {
        throw const MudbaseException(
          'The server did not return the created document.',
          0,
        );
      }
      return document;
    } on DioException catch (error) {
      throw MudbaseException.fromDioException(error);
    }
  }

  Future<Map<String, dynamic>> update(
    String collectionId,
    String documentId,
    Map<String, dynamic> body, {
    required String token,
  }) async {
    try {
      final response = await _sdk.dio.patch<dynamic>(
        _basePath(collectionId, documentId),
        data: body,
        options: _authOptions(token, contentType: 'application/json'),
      );
      final document = _unwrapDocument(response);
      if (document == null) {
        throw const MudbaseException(
          'The server did not return the updated document.',
          0,
        );
      }
      return document;
    } on DioException catch (error) {
      throw MudbaseException.fromDioException(error);
    }
  }

  Future<void> delete(
    String collectionId,
    String documentId, {
    required String token,
  }) async {
    try {
      await _sdk.dio.delete<dynamic>(
        _basePath(collectionId, documentId),
        options: _authOptions(token),
      );
    } on DioException catch (error) {
      throw MudbaseException.fromDioException(error);
    }
  }

  Map<String, dynamic>? _unwrapDocument(Response<dynamic> response) {
    final body = response.data;
    if (body is! Map<String, dynamic>) return null;
    final data = body['data'];
    return data is Map<String, dynamic> ? data : null;
  }
}
